rust/src/index/vector/pq.rs

```diff
@@ -257,14 +276,68 @@ impl VectorIndex for PQIndex {
         let row_ids =
             read_fixed_stride_array(reader, &DataType::UInt64, row_id_offset, length, ..).await?;
 
+        // Precollecting everything to a hash map requires async closure. This way we avoid that
+        let mut deletion_vec_cache: HashMap<u64, DeletionVector> = HashMap::new();
+
+        let mut filtered_row_id_builder = UInt64Builder::new();
+        // TODO: consider a more optimized way of reading
+        // group by frag_id and check per frag in one go
+        for row_id in as_primitive_array::<UInt64Type>(row_ids.as_ref()) {
+            let row_id = row_id.unwrap();
+            let frag_id = row_id >> 32;
+            let local_row_id = row_id as u32;
+            if !deletion_vec_cache.contains_key(&frag_id) {
+                let fragment = self
+                    .manifest
+                    .as_ref()
+                    .fragments
+                    .as_ref()
+                    .iter()
+                    .find(|frag| frag.id == frag_id);
+                let deletion_vec = match fragment {
+                    Some(frag) => {
+                        let dvec = read_deletion_file(
+                            self.object_store.as_ref().base_path(),
+                            frag,
+                            self.object_store.as_ref(),
+                        )
+                        .await
+                        .map(|dvec_option| dvec_option.unwrap_or(DeletionVector::NoDeletions));
+                        if dvec.is_err() {
+                            return Err(Error::Index {
+                                message: format!(
+                                    "Failed to read deletion file for fragment {}",
+                                    frag_id
+                                ),
+                            });
+                        }
+                        dvec.unwrap()
```

Comment:
This logic masks the underlying error.

Could we just do:

```suggestion
                        dvec?
```